[BoS] Add `stack.bind` and remove `stack.new_switch`

tlively opened the pull request:
We previously allowed `stack.new` to bind arguments to the newly allocated
stack. However, producers might just as well want to perform partial application
with any stack reference, not just with newly allocated stacks. To simplify the
typing and semantics of `stack.new` and allow partial application to be used
more broadly, remove all operands of `stack.new` and add a separate `stack.bind`
instruction to perform partial application.

Since `stack.new` no longer does partial application, the optimization
opportunity realizable by a combined `stack.new_switch` instruction over a
sequence of `(stack.new) (switch)` is much smaller than it was before. Since
`stack.new_switch` no longer has as much benefit, remove it from the proposed
instruction set.

Update the prose and examples to replace uses of `stack.new_switch` with uses of
`stack.new` combined with `switch`.

tlively commented:
* **#56** <a href="https://app.graphite.dev/github/pr/WebAssembly/stack-switching/56?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @tlively and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

tlively commented:
friendly ping @fgmccabe

fgmccabe commented on proposals/bag-o-stacks/Explainer.md:
```diff
@@ -134,9 +134,9 @@ absheaptype ::= ... | stack | nostack
 ### Life-cycle of a coroutine
 
-A coroutine is started using the `stack.new_switch` instruction. This performs the equivalent of a function call – on a new stack resource. In addition to the arguments normally expected in a function call, an additional argument is provided that is a stack reference to the caller code -- the caller is suspended as a result of the `stack.new_switch` instruction.
+A coroutine is allocated in the suspended state using the `stack.new` instruction. The initial `switch` to the newly allocated coroutine performs the equivalent of a function call on the new stack resource. In addition to the arguments provided to the `switch`, an additional argument is provided that is a stack reference to the caller code -- the caller is suspended as a result of the `switch` in ...
```
I remain uncomfortable with this change:
in particular, stack.bind cannot completely replace the stack.new_switch instruction.

For the sake of argument, lets call the recursive group of types P (for protocol).

Then, clearly

stack.switch x |= x in P

but also,

C|- stack.new f |= f expands to func x [] and x in P

and

C|- stack.bind x y |= x in P and y in P

This is because of the common rt param in all the stack types.

This effectively eliminates one of the main use cases of stack.new_switch - of binding free variables to a coroutine function before entry into the coroutine proper.

In addition, stack.bind seems less powerful and less useful than cont.bind; again, because you cannot leave (or enter) the recursion group P.

tlively commented:
@fgmccabe, I updated the generator example to show how the generator can be initialized with or without `stack.bind`. To keep the example as simple as possible, I removed the unused cancellation flag and use the null stack reference from `switch_retire` to signal the end.

fgmccabe commented on proposals/bag-o-stacks/Explainer.md:
```diff
@@ -289,93 +278,125 @@ where $genCmd has a single i32 which contains the command to the generator, and
-Our example code handles the command code by a `br_table` instruction that either continues to the next block or arranges to exit the entire function.
+Whenever a `switch` instruction is used, it must be followed by instructions that store the return stack reference and use any sent values. In this example, the return stack reference is stored in the `$toConsumer` local variable, replacing its previous value which is no longer valid.
 
 >There is one aspect of building a generator that is not addressed by our code so far: how to start it. We will look at this in more detail as we look at the consumer side of yield-style generators next.
```
Should delete this line now.

fgmccabe reviewed (approved):
Looks good.
